pkg/controller/multiarchbuildconfig/multiarchbuildconfig.go

```diff
@@ -0,0 +1,361 @@
+
+	// TODO: Now we hardcode the URL for the heterogeneous cluster.
+	// TODO: We need to push it to the app.ci registry or quay.io.
+	registryURL = "registry.multi-build01.arm-build.devcluster.openshift.com"
+)
+
+func AddToManager(mgr manager.Manager, buildClient BuildClient, architectures []string) error {
+	c, err := controller.New(controllerName, mgr, controller.Options{
+		MaxConcurrentReconciles: 1,
+		Reconciler: &reconciler{
+			logger:        logrus.WithField("controller", controllerName),
+			client:        mgr.GetClient(),
+			architectures: architectures,
+			buildClient:   buildClient,
+		},
+	})
+	if err != nil {
+		return fmt.Errorf("failed to construct controller: %w", err)
+	}
+
+	predicateFuncs := predicate.Funcs{
+		CreateFunc:  func(e event.CreateEvent) bool { return true },
+		DeleteFunc:  func(event.DeleteEvent) bool { return false },
+		UpdateFunc:  func(event.UpdateEvent) bool { return false },
+		GenericFunc: func(event.GenericEvent) bool { return false },
+	}
+	if err := c.Watch(source.Kind(mgr.GetCache(), &v1.MultiArchBuildConfig{}), mabcHandler(), predicateFuncs); err != nil {
+		return fmt.Errorf("failed to create watch for MultiArchBuildConfig: %w", err)
+	}
+
+	return nil
+}
+
+func mabcHandler() handler.EventHandler {
+	return handler.EnqueueRequestsFromMapFunc(func(ctx context.Context, o ctrlruntimeclient.Object) []reconcile.Request {
+		mabc, ok := o.(*v1.MultiArchBuildConfig)
+		if !ok {
+			logrus.WithField("type", fmt.Sprintf("%T", o)).Error("Got object that was not a MultiArchBuildConfig")
+			return nil
+		}
+
+		return []reconcile.Request{
+			{NamespacedName: types.NamespacedName{Namespace: mabc.Namespace, Name: mabc.Name}},
+		}
+	})
+}
+
+type reconciler struct {
+	logger        *logrus.Entry
+	client        ctrlruntimeclient.Client
+	buildClient   BuildClient
+	architectures []string
+}
+
+func (r *reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	logger := r.logger.WithField("request", req.String())
+	err := r.reconcile(ctx, req, logger)
+	if err != nil {
+		logger.WithError(err).Error("Reconciliation failed")
+	} else {
+		logger.Info("Finished reconciliation")
+	}
+	return reconcile.Result{}, controllerutil.SwallowIfTerminal(err)
+}
+
+func (r *reconciler) reconcile(ctx context.Context, req reconcile.Request, logger *logrus.Entry) error {
+	logger = logger.WithField("multiarchbuildconfig_name", req.Name)
+	logger.Info("Starting reconciliation")
+
+	mabc := &v1.MultiArchBuildConfig{}
+	if err := r.client.Get(ctx, req.NamespacedName, mabc); err != nil {
+		return fmt.Errorf("failed to get the MultiArchBuildConfig: %w", err)
+	}
+
+	if mabc.Status.State == v1.SuccessState || mabc.Status.State == v1.FailureState {
+		return nil
+	}
+
+	if err := r.handleMultiArchBuildConfig(ctx, mabc); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (r *reconciler) handleMultiArchBuildConfig(ctx context.Context, mabc *v1.MultiArchBuildConfig) error {
```

Comment:
This function is not executed in any of tests.

Reply:
Yes, but all of the main operations inside are tested.